### ISSUE TITLE
[Feature] Integration Tests using docker-compose

### ISSUE THREAD

**Issue by gandarez** [fluentmigrator/fluentmigrator#1131]
Hi there!

What you guys think of creating a real integration tests using `docker-compose`? We can easily attach as many databases we want and then run the tests in a `live` environment.
I created a `poc` to show how it should work. On the terminal simply run `docker-compose up` to get the tests running.

The repository is here:
https://github.com/gandarez/fluentmigrator/tree/feature/integration-tests

**Comment by fubar-coder** [comment]
Related to #1129, #1067, and #1061.

**Comment by gandarez** [comment]
Let's move this on! I haven't seem any evolution on those issues/feature requests. I think using `docker dind` in Azure DevOps is definetely possible and since this is possible we can write our own `tasks` instead of using the predefined ones from the pipeline.

**Comment by jzabroski** [comment]
@gandarez Thanks for your help.  I have wanted to make progress on this but I take a lot of time just responding to issues for this project.  I've started closing issues that aren't actionable in an effort to get back to writing code vs. just providing support.

I also never heard of docker dind.  That is really cool, and thanks for sharing that.  I just gave a talk on Docker at my user group, New England Microsoft Developers (https://www.meetup.com/NE-MSFT-Devs/events/266758838/) but I'm relatively a Docker newbie.

**Comment by jzabroski** [comment]
@gandarez  So is baby step 1 proving that we can execute `docker run dind` in AzureDevOps?

Also, innocent question:

> What you guys think of creating a real integration tests using `docker-compose`? We can easily attach as many databases we want and then run the tests in a `live` environment.

Why did you put `live` in code text?  It's not a docker keyword I'm familiar with. Can you teach me what this refers to?

In looking around (also never having heard of docker dind), it seems running builds through dind is a practice adopted by the GitLab CI Community?

**Comment by jzabroski** [comment]
@gandarez Hi Carlos, I did some reading on my lunch break.  It seems, at least as of 2018, the recommendation is to not run Docker In Docker for Continuous Integration: https://itnext.io/docker-in-docker-521958d34efd

Perhaps you're more experienced than the person writing this article, and perhaps you've written your own blog articles on doing docker-in-docker and feel differently about this than the author I linked to.  However, digging down, the guy saying "Think Again" is Jerome Pettazoni, the very guy who added the `--privileged` flag to Docker, so I'm curious how you will out-smart the guy who started PipeWork and made contributions to Docker... I'm all for learning from you and this is rather exciting development for FluentMigrator.

**Comment by jzabroski** [comment]
@gandarez Hi Carlos, I thought about this a bit further.  I think I understand what you're trying to do.

1. You want to build the code inside a docker container.
2. You want to stand up database management systems inside docker containers, one container per database.
3. You want to run the tests inside a docker container.

Between steps 1 and 3, you want to stand up databases inside a docker container, and allow those tests to connect to well known ports inside the docker container.  To orchestrate 1, 2 and 3 happen parsimoniously, you want to use docker compose.

However, you don't know what ports are used by the operating system at the time your running these tests, so to maximize portability and reproducibility across anyone trying to contribute to this project, you want to stand up 1 and 2 and 3 _inside docker itself_ and thereby allow anyone using Docker to do so for developing FluentMigrator while not screwing with whatever other ports they're using.

If I got all that right, then great - I think we're starting to get on the same page.
There are probably a few minor gotchas to work out with this approach:

1. Can we avoid requiring `--privileged` flag and instead use `--read-only` somehow?
2. Do we need a bootstrap script to allow users to configure their firewall rules?  My guess is, if everything runs inside docker, they don't need any firewall rules?  Is that part of the genius behind your approach?

**Comment by gandarez** [comment]
Hi @jzabroski 

The `docker-compose` is only to do the integration tests (or maybe we can also run the unit ones, why not?). Using by this way we isolate to docker context (docker concept) and then we easily run every single test against a real (which I called `live`) database. As you can see in the `docker-compose.yml` file we can create a bunch of database running all at the same time and then we just need to wait the execution of the integration tests. Have you had a chance to try it out? It is simple as running `docker-compose up` in the terminal and wait for the results.

Answering your questions
```
Can we avoid requiring --privileged flag and instead use --read-only somehow?
```
I don't know what you're talking about.

```
Do we need a bootstrap script to allow users to configure their firewall rules? My guess is, if everything runs inside docker, they don't need any firewall rules? Is that part of the genius behind your approach?
```
There's no extra setup needed in any machine to run `docker-compose`. At least, and of course, it needs `docker` and `docker-compose` installed.

**Comment by jzabroski** [comment]
I kind of was impatient and started asking around about best practices.  https://www.reddit.com/r/docker/comments/e8fgim/is_it_safe_to_mount_the_host_machines_docker/faccnpt/

> 
> > Can we avoid requiring `--privileged` flag and instead use `--read-only` somehow?
> 
> I don't know what you're talking about.

When you execute `docker run`, you can specify [`--privileged`](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities) flag to enable/disable whether the container can run as a privileged process.  In Jerome Petazonni's blog post [Using Docker-in-Docker for your CI or testing environment? Think twice.](https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/), he mentions he "contributed the -privileged flag in Docker and wrote the first version of dind."  I'd say he is somewhat of an expert.

You can also specify `[--read-only](https://docs.docker.com/engine/reference/commandline/run/#mount-volume--v---read-only)`, although, today, I discovered that --read-only docker daemon socket can create a child container, so --read-only doesn't offer much in the way of security.

tl;dr: Don't let me slow you down, but I am just trying to play catch up with you in understanding some of this.  You might be more gung ho than me; I tend to go very deep and wide into things before making much progress.

> As you can see in the `docker-compose.yml` file we can create a bunch of database running all at the same time and then we just need to wait the execution of the integration tests. Have you had a chance to try it out?

Sorry - I didn't think you were fully done yet - I see now the docker-compose.yml is way below the Dockerfile due to lexicographic sorting.

**Comment by jzabroski** [comment]
Thanks for all your help.  I will give this a shot. Hope to learn a lot.

**Comment by gandarez** [comment]
What about if we create a slack channel for this project and start discussion there?

**Comment by jzabroski** [comment]
We have a gitter channel

**Comment by jzabroski** [comment]
@gandarez  Here is the link - it's also in the readme: https://gitter.im/FluentMigrator/fluentmigrator

Gitter uses your github username so its easy to find me and Mark (@fubar-coder ) and @eloekset (the one who helped us set up AzureDevOps and get off TravisCI).

**Comment by jzabroski** [comment]
@gandarez  Am I calling this wrong?  I'm still a docker newbie.

```powershell
docker build https://github.com/gandarez/fluentmigrator.git#feature/integration-tests -t fluentmigrator --progress tty
```
```
time="2019-12-16T10:11:30-05:00" level=error msg="failed to dial gRPC: unable to upgrade to h2c, received 500"
time="2019-12-16T10:11:30-05:00" level=error msg="Can't add file \\\\?\\C:\\Users\\john.zabroski\\AppData\\Local\\Temp\\docker-build-git564232019\\azure-pipelines.yml to tar: io: read/write on closed pipe"
time="2019-12-16T10:11:30-05:00" level=error msg="Can't close tar writer: io: read/write on closed pipe"
error during connect: Post http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.40/build?buildargs=%7B%7D&cachefrom=%5B%5D&cgroupparent=&cpuperiod=0&cpuquota=0&cpusetcpus=&cpusetmems=&cpushares=0&dockerfile=Dockerfile&labels=%7B%7D&memory=0&memswap=0&networkmode=default&rm=1&session=ml0q17fgluhdcyss0jwb0xxun&shmsize=0&t=fluentmigrator&target=&ulimits=null&version=1: context canceled
```

**Comment by gandarez** [comment]
Yes, it's **not** correct. Clone my repo, open a terminal and then simply call `docker-compose build` or `docker-compose up` depending on your needs.

**Comment by jzabroski** [comment]
Thanks, I got it working.  Just need to figure out how to inspect the test output xml?  Also, how would we debug this?  Thanks for your help.

```powershell
 git clone https://github.com/gandarez/fluentmigrator.git FluentMigratorDocker
cd FluentMigratorDocker
git checkout feature/integration-tests
docker-compose build
```

**Comment by gandarez** [comment]
1. To inspect the output I need to configure that to expose that file outside the container. Will do that.
2. There's no way to debug that since you running a deployed container. To debug you must go to VS stuff and run from there. But to do that there's a trick to get it working.

**Comment by jzabroski** [comment]
What's the trick?  I read about it last night and it seems Microsoft's DockerTools project had support for it briefly but then removed it from Visual Studio 15.8 and up.  It sounds like VSCode (lil bro Visual Studio) supports debugging code running inside a docker container.

This article covers some stuff here: https://www.aaron-powell.com/posts/2019-04-04-debugging-dotnet-in-docker-with-vscode/
This article covers debugging ASP.NET Core apps in Docker with Visual Studio 2019 (big bro): https://docs.microsoft.com/en-us/visualstudio/containers/edit-and-refresh?view=vs-2019 - Toward the bottom of this article, they discuss Console Apps, which is what FM is.  It sounds like it isn't supported, BUT if we create a FluentMigrator.dcproj in the solution, it will just automagically work assuming we dont do anything too weird.

**Comment by jzabroski** [comment]
The last thing seems to be that we should modify the build.targets so that PreBuild event clearn out the old docker container.  I don't see anyone talk about it but in the edit and refresh article I linked, they say the only way to trigger a new container is to modify the Dockerfile.  https://docs.microsoft.com/en-us/visualstudio/containers/edit-and-refresh?view=vs-2019#container-reuse

**Comment by gandarez** [comment]
Keep in mind that the `Docker` proposal here in only to run the integration tests which has no mocked data in it. Using the power of `docker-compose` we can achieve that. I'm not talkign about `MSBUILD`, `PreBuild`, nuget deployment, etc.

**Comment by jzabroski** [comment]
I understand, but what will end up happening is people contribute fixes, and they dont run the integration tests locally, and the build fails on azuredevops, and they will want to know how they can debug the failed tests if it requires running docker-compose with 8 database engines.

Maybe I'm trying to bite off too much at once.  Overall, I like where we're heading

**Comment by gandarez** [comment]
Ok I got it. Firstly, as a developer you must run the integration tests locally before publishing to any branch or doing a PR. Secondly, we could easily attach the integration tests using docker-compose into the pipeline, if it fails you break the execution of it. Finally, we end up with a solid and realiable test for all (maybe not all) database that FM supports.

My proposal is simply not mocking anything and not depending upon a dev machine to run the tests.

**Comment by jzabroski** [comment]
I guess we could have SqlServer.Dockerfile, etc. and a bunch of smaller docker-compose.yml files that only run dotnet test on a specific subset of XUnit Traits https://docs.microsoft.com/en-us/dotnet/core/testing/selective-unit-tests

**Comment by jzabroski** [comment]
@gandarez Separately, are you familiar with the *.dcproj extension in Visual Studio? https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/docker/visual-studio-tools-for-docker?view=aspnetcore-3.1#docker-compose

This would allow some additional flexibility in running some of these tests on both Linux and Windows.  I believe the dcproj is a decent fiddling layer/abstraction that allows using MSBuild to help orchestrate the container setup.

**Comment by jzabroski** [comment]
@gandarez So, I'm done researching this.

I think we should use your approach for now, and as testcontainers-dotnet gets more mature, start to move away from docker-compose to run integration tests.  Given MySQL is the buggiest platform we seem to support right now, I think it's good enough if we just get integration tests running on MySQL and suss out the open issues over the next 3-6 months.

*.dcproj might be a classic Microsoft mistake, in that they're basically trying to build a meta-framework around an "MVP" tool like docker-compose that has never graduated to full-fledged project status.  It has all the smell of DACPACs. We probably shouldn't tie FluentMigrator to *.dcproj, as we'll eventually be sorry (just my hunch).

**Comment by PhenX** [comment]
I think this issue can be closed now, @jzabroski ?
